### ISSUE TITLE
Issue with using list[SomeProtocolType]

### ISSUE THREAD

**Issue by ericroberts** [microsoft/pyright#4762]
**Describe the bug**
When using a type annotation like `list[SomeProtocolType]` pyright sometimes reports that things that are compatible are incompatible.

**To Reproduce**

Run pyright against this code:

``` python
# pyright: strict

from typing import Protocol

class MyProtocol(Protocol):
    x: int

class ImplementsMyProtocol:
    x: int

def x(arg: list[MyProtocol]):
    ...

def y(arg: list[ImplementsMyProtocol]):
    # Argument of type "list[ImplementsMyProtocol]" cannot be assigned
    # to parameter "arg" of type "list[MyProtocol]" in function "x"
    x(arg)

    # Argument of type "list[ImplementsMyProtocol]" cannot be assigned
    # to parameter "arg" of type "list[MyProtocol]" in function "x"
    y = [ImplementsMyProtocol()]
    x(y)

    # no error
    x([ImplementsMyProtocol()])
```

**Expected behavior**
I believe this code should have no errors.

**Screenshots or Code**

Here's a screenshot of the output which should be reproducible with the code above.

<img width="761" alt="image" src="https://user-images.githubusercontent.com/695547/224429015-4b0c80d3-d752-4ba2-88de-b238c4a65ee7.png">

**VS Code extension or command-line**

Command line, version 1.1.298

**Comment by erictraut** [comment]
Pyright is correct here. The type parameter to the `list` class is invariant, so the type must match exactly. Subtypes (structural or nominal) are not allowed. For more information about why this must be the case, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts?id=type-assignability).

Not surpisingly, mypy generates the same error as pyright here.

This will type check without errors:
```python
# pyright: strict
from typing import Protocol

class MyProtocol(Protocol):
    x: int

class ImplementsMyProtocol:
    x: int

def x(arg: list[MyProtocol]): ...

def y(arg: list[MyProtocol]):
    x(arg)

    y: list[MyProtocol] = [ImplementsMyProtocol()]
    x(y)

    x([ImplementsMyProtocol()])
```

**Comment by ericroberts** [comment]
I appreciate you taking the time to respond with an explanation to something you must have explained several times before! The documentation was very helpful and I was able to find a way around my issue. Thank you!
